azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -36,17 +43,178 @@ def __init__(self, callback):
         Initializer for PipelineOperation objects.
 
         :param Function callback: The function that gets called when this operation is complete or has
-          failed.  The callback function must accept A PipelineOperation object which indicates
-          the specific operation which has completed or failed.
+            failed. The callback function must accept A PipelineOperation object which indicates
+            the specific operation which has completed or failed.
         """
         if self.__class__ == PipelineOperation:
             raise TypeError(
                 "Cannot instantiate PipelineOperation object.  You need to use a derived class"
             )
         self.name = self.__class__.__name__
-        self.callback = callback
+        self.callback_stack = []
         self.needs_connection = False
-        self.completed = False
+        self.completed = False  # Operation has been fully completed
+        self.completing = False  # Operation is in the process of completing
+        self.error = None  # Error associated with Operation completion
+
```

Comment:
Something to think about: These 3 attributes (completed, completing, and error) are all part of one thing.  I hesitate to use the word state, but these combine together to form the state of the op, and not all combinations are valid.  For states, I think we have: not_complete, completing_with_error, completing_without_error, completed_with_error, and completed_without_error.  With these 3 attributes, there's 8 possible combinations, but only 5 valid states.  This worries me, especially as code grows and the possibility of internally inconsistent ops gets bigger.

Reply:
Yeah I'm a little concerned about this as well - for now I feel safe with the value checks I put on each operation, but it's not an ideal situation. As I was working on this class, some of the functionality started to resemble a state machine a bit too much for my liking.

---
In reply to: [347141190](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347141190) [](ancestors = 347141190)